### PR TITLE
GH-43414: [C++][Compute] Fix invalid memory access when resizing var-length buffer in row table

### PR DESCRIPTION
**Pull request by zanmato1984**
### Rationale for this change

As #43414 explained. The UT in PR reproduces this issue well (may need ASAN to detect).

### What changes are included in this PR?

Check if `is_fixed_length` before treating the second buffer as offset.

### Are these changes tested?

UT included.

### Are there any user-facing changes?

None.

* GitHub Issue: #43414

### CODE REVIEW

**Comment by zanmato1984** [comment]
Hi @pitrou @westonpace , would you please take a look? Thanks.

**Comment by zanmato1984** [comment]
Looping in @felipecrv as well. Would you help to take a look? Thanks.

File: cpp/src/arrow/compute/row/row_internal.cc [hunk 1]
```diff
@@ -293,8 +293,12 @@ Status RowTableImpl::ResizeFixedLengthBuffers(int64_t num_extra_rows) {
 }
 
 Status RowTableImpl::ResizeOptionalVaryingLengthBuffer(int64_t num_extra_bytes) {
+  if (metadata_.is_fixed_length) {
+    return Status::OK();
+  }
```
Comment (felipecrv) [inline cpp/src/arrow/compute/row/row_internal.cc:298]: IMO it's more elegant to document that `ResizeOptional*VaryingLength*Buffer` has a pre-condition

`/// \pre !metadata_.is_fixed_length`

Then you put a `DCHECK(!metadata_.is_fixed_length)` and review the 3 callsites to ensure this is only called when the pre-condition is respected.

Comment (felipecrv) [reply cpp/src/arrow/compute/row/row_internal.cc:298]: In the first callsite it's clearly inside a branch where `!metadata_.is_fixed_length`.

The second callsite should have an if guarding the call to `ResizeOptionalVaryingLengthBuffer` as it doesn't make sense to resize varying length buffers if there aren't any. The client code become more explicit.

```diff
 Status RowTableImpl::AppendEmpty(uint32_t num_rows_to_append,
                                  uint32_t num_extra_bytes_to_append) {
   RETURN_NOT_OK(ResizeFixedLengthBuffers(num_rows_to_append));
-  RETURN_NOT_OK(ResizeOptionalVaryingLengthBuffer(num_extra_bytes_to_append));
+  if (!metadata_.is_fixed_length) {
+    RETURN_NOT_OK(ResizeOptionalVaryingLengthBuffer(num_extra_bytes_to_append));
+  }
   num_rows_ += num_rows_to_append;
   if (metadata_.row_alignment > 1 || metadata_.string_alignment > 1) {
     memset(rows_->mutable_data(), 0, bytes_capacity_);
   }
   return Status::OK();
 }
```

Comment (zanmato1984) [reply cpp/src/arrow/compute/row/row_internal.cc]: Thank you! Addressed.

File: cpp/src/arrow/compute/row/row_internal.h [hunk 2]
```diff
@@ -220,7 +220,14 @@ class ARROW_EXPORT RowTableImpl {
   }
 
  private:
+  /// @brief Resize the fixed length buffers to store `num_extra_rows` more rows. The
+  /// fixed length buffers are buffers_[0] for null masks, buffers_[1] for row data if the
+  /// row is fixed length, or for row offsets otherwise.
   Status ResizeFixedLengthBuffers(int64_t num_extra_rows);
+
+  /// @brief Resize the optional varying length buffer to store `num_extra_bytes` more
```
Comment (felipecrv) [inline cpp/src/arrow/compute/row/row_internal.h:228]:
```suggestion
  /// \brief Resize the optional varying length buffer to store `num_extra_bytes` more
```

**Review by felipecrv — APPROVED** [review]

**Comment by zanmato1984** [comment]
The CI failure seems unrelated. Shall we move on with this? Thanks.

File: cpp/src/arrow/compute/row/row_internal.h (hunk 2 above)
Comment (zanmato1984) [reply cpp/src/arrow/compute/row/row_internal.h]: Thanks for correcting this!

**Pull request merged**